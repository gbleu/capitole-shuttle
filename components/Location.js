import React, { Component } from 'react';
import { Platform, Text, View, StyleSheet } from 'react-native';
import { Constants, Location, Permissions } from 'expo';
import moment from 'moment';

import services from '../services';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    // Invariant Violation: Invalid prop `paddingTop` supplied to `StyleSheet container`.
    // paddingTop: Constants.statusBarHeight,
    paddingTop: 20,
    backgroundColor: '#ecf0f1'
  },
  paragraph: {
    margin: 24,
    fontSize: 26,
    textAlign: 'center'
  }
});

export default class App extends Component {
  state = {
    location: null,
    sub: null,
    origin: {},
    departures: [null, null, null],
    traffic: null,
    errorMessage: null
  };

  componentWillMount() {
    if (Platform.OS === 'android' && !Constants.isDevice) {
      this.setState({
        errorMessage: 'Oops, this will not work on Sketch in an Android emulator. Try it on your device!'
      });
    } else {
      // get initial location
      this.getLocation();
      // watch location
      this.watchLocation();
      // refresh departures every minute
      this.interval = setInterval(this.updateDepartures, 1 * 60 * 1000);
    }
  }

  componentWillUnmount() {
    clearInterval(this.interval);

    if (this.state.sub) {
      this.state.sub.remove();
    }
  }

  getLocation = async () => {
    const { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status !== 'granted') {
      this.setState({
        errorMessage: 'Permission to access location was denied'
      });
      return;
    }

    console.log('get current location');
    const location = await Location.getCurrentPositionAsync({});
    await this.updateLocation(location);
  };

  watchLocation = async () => {
    console.log('watch current location');
    const sub = Location.watchPositionAsync({}, (location) => {
      console.log('new location received');
      this.updateLocation(location);
    });
    this.setState({ sub });
  };

  updateLocation = async (location) => {
    console.log('update location', JSON.stringify(location));

    const origin = services.guessOrigin(location.coords);
    console.log('origin', JSON.stringify(origin));

    this.setState({ location, origin });
    console.log('location updated');

    await this.updateDepartures();
  };

  updateDepartures = async () => {
    const { origin } = this.state;

    const departures = services.getNextDepartures(origin.key);
    console.log('departures', JSON.stringify(departures));

    const traffic = await services.trafficTo(origin.key);
    console.log(`traffic +${traffic} min.`);

    this.setState({ departures, traffic });
    console.log('departures updated');
  };

  format = (date) => {
    if (!date) return '--:--';

    return date && moment(date).format('HH[h]mm');
  };

  render() {
    let text = 'acquiring location ...';
    if (this.state.errorMessage) {
      text = this.state.errorMessage;
    } else if (this.state.origin.key) {
      if (this.state.origin.key === 'defense') {
        text = 'defense -> capitole';
      } else {
        text = 'capitole -> defense';
      }
    }

    return (
      <View style={styles.container}>
        <Text style={styles.paragraph}>{text}</Text>
        <View style={{ flexDirection: 'row' }}>{
          // eslint-disable-next-line react/no-array-index-key
          this.state.departures.map((it, i) => <Text key={`departure-${i}`} style={styles.paragraph}>{this.format(it)}</Text>)
        }</View>
        {this.state.traffic && <Text style={styles.paragraph}>+{this.state.traffic} min.</Text>}
      </View>
    );
  }
}
