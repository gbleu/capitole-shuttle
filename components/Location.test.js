/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';

import Location from './Location';

describe('Location component', () => {
  it('renders as expected', async () => {
    const wrapper = shallow(<Location />);
    expect(wrapper).toMatchSnapshot();

    wrapper.setState({ errorMessage: 'errorMessage' });
    expect(wrapper).toMatchSnapshot();

    wrapper.setState({ errorMessage: null, origin: { key: 'defense' } });
    expect(wrapper).toMatchSnapshot();

    wrapper.setState({ errorMessage: null, origin: { key: 'capitole' } });
    expect(wrapper).toMatchSnapshot();

    wrapper.setState({ departures: [new Date(0, 0, 0, 0, 0, 0, 0), null, null], traffic: 2 });
    expect(wrapper).toMatchSnapshot();

    const remove = jest.fn();
    wrapper.setState({ sub: { remove } });
    wrapper.unmount();
    expect(wrapper).toMatchSnapshot();
    expect(remove).toBeCalled();
  });
});
