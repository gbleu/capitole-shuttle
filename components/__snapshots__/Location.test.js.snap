// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location component renders as expected 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 20,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 26,
        "margin": 24,
        "textAlign": "center",
      }
    }
  >
    acquiring location ...
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
  </View>
</View>
`;

exports[`Location component renders as expected 2`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 20,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 26,
        "margin": 24,
        "textAlign": "center",
      }
    }
  >
    errorMessage
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
  </View>
</View>
`;

exports[`Location component renders as expected 3`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 20,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 26,
        "margin": 24,
        "textAlign": "center",
      }
    }
  >
    defense -&gt; capitole
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
  </View>
</View>
`;

exports[`Location component renders as expected 4`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 20,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 26,
        "margin": 24,
        "textAlign": "center",
      }
    }
  >
    capitole -&gt; defense
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
  </View>
</View>
`;

exports[`Location component renders as expected 5`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ecf0f1",
      "flex": 1,
      "justifyContent": "center",
      "paddingTop": 20,
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 26,
        "margin": 24,
        "textAlign": "center",
      }
    }
  >
    capitole -&gt; defense
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      00h00
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 26,
          "margin": 24,
          "textAlign": "center",
        }
      }
    >
      --:--
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 26,
        "margin": 24,
        "textAlign": "center",
      }
    }
  >
    +
    2
     min.
  </Text>
</View>
`;

exports[`Location component unmount as expected 1`] = `""`;
